File: assets/js/hooks/canvas_toolbar.js
/**
 * CanvasToolbar — unified Phoenix LiveView Hook for canvas floating toolbars.
 *
 * Repositions the floating toolbar after LiveView patches update its content.
 * Works with both flow (Rete) and scene (Leaflet) canvases via data-canvas-id.
 *
 * Usage: <div phx-hook="CanvasToolbar" data-canvas-id="flow-canvas">
 */
export const CanvasToolbar = {
  updated() {
    const canvasId = this.el.dataset.canvasId;
    if (!canvasId) return;

    const canvas = document.getElementById(canvasId);
    if (canvas?.__floatingToolbar) {
      canvas.__floatingToolbar.reposition();
    }
  },
};

File: lib/storyarn_web/components/canvas_toolbar.ex
defmodule StoryarnWeb.Components.CanvasToolbar do
  @moduledoc """
  Shared floating toolbar wrapper for canvas editors (flow, scene).

  Handles the outer positioning container, visibility toggling via CSS class,
  and the CanvasToolbar LiveView hook for repositioning after content patches.
  """

  use Phoenix.Component

  attr :id, :string, required: true
  attr :canvas_id, :string, required: true
  attr :visible, :boolean, default: false
  attr :z_class, :string, default: "z-30"

  slot :inner_block, required: true

  def canvas_toolbar(assigns) do
    ~H"""
    <div
      id={@id}
      phx-hook="CanvasToolbar"
      data-canvas-id={@canvas_id}
      class={["absolute canvas-toolbar", @z_class]}
    >
      <div :if={@visible} class="floating-toolbar">
        {render_slot(@inner_block)}
      </div>
    </div>
    """
  end
end
